File: scripts/sovereign-backend/agent_runtime/neuro_architecture_contract.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
import hashlib
import json
import re
from typing import Any, Mapping, Sequence

_SHA40 = re.compile(r"^[0-9a-f]{40}$")
_SHA256 = re.compile(r"^[0-9a-f]{64}$")
_ID = re.compile(r"^[a-z0-9][a-z0-9._:-]{1,159}$")


class EvidenceClass(str, Enum):
    METAPHOR = "E0"
    PLAUSIBLE_ANALOGY = "E1"
    BROAD_FUNCTIONAL_EVIDENCE = "E2"
    ANATOMICAL_CONNECTION_EVIDENCE = "E3"
    REPRODUCIBLE_FORMAL_PROPERTY = "E4"


class Lane(str, Enum):
    SENSORY_INTAKE = "sensory-intake"
    THALAMIC_ROUTING = "thalamic-routing"
    REFLEX_SAFETY = "reflex-safety"
    DETERMINISTIC_VERIFICATION = "deterministic-verification"
    COGNITIVE_SIDE_CHANNEL = "cognitive-side-channel"
    EVIDENCE = "evidence"
    PERSISTENCE = "persistence"
    CEREBELLAR_CORRECTION = "cerebellar-correction"
    MOTOR_AUTHORIZATION = "motor-authorization"
    HOMEOSTASIS = "homeostasis"
    QUARANTINE = "quarantine"


CANONICAL_TRUTH_LANES = frozenset({
    Lane.DETERMINISTIC_VERIFICATION,
    Lane.EVIDENCE,
    Lane.MOTOR_AUTHORIZATION,
})

NON_CANONICAL_LANES = frozenset({Lane.COGNITIVE_SIDE_CHANNEL})


@dataclass(frozen=True, slots=True)
class NeuroAliasBinding:
    alias: str
    canonical_component: str
    evidence_class: EvidenceClass
    software_scope: str
    scientific_claim: str
    limitations: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        if not self.alias.strip():
            raise ValueError("alias must not be empty")
        if not _ID.fullmatch(self.canonical_component):
            raise ValueError("canonical_component must be a stable lowercase identifier")
        if self.software_scope not in {"sovereign-studio-ato", "arelorian-wasd-side-channel"}:
            raise ValueError("software_scope violates project isolation")
        if not self.scientific_claim.strip():
            raise ValueError("scientific_claim must not be empty")


@dataclass(frozen=True, slots=True)
class EvidenceEnvelope:
    schema_version: str
    system_id: str
    revision_sha: str
    policy_sha256: str
    event_id: str
    lane: Lane
    tick: int
    sequence: int
    payload_sha256: str
    causal_parent_sha256: str
    previous_evidence_sha256: str
    producer_identity: str
    canonical: bool
    side_channel_reference: str = ""

    def __post_init__(self) -> None:
        if self.schema_version != "sovereign.neuro-architecture-envelope.v1":
            raise ValueError("unsupported schema_version")
        if not _ID.fullmatch(self.system_id):
            raise ValueError("invalid system_id")
        if not _SHA40.fullmatch(self.revision_sha):
            raise ValueError("invalid revision_sha")
        for value, name in (
            (self.policy_sha256, "policy_sha256"),
            (self.payload_sha256, "payload_sha256"),
            (self.causal_parent_sha256, "causal_parent_sha256"),
            (self.previous_evidence_sha256, "previous_evidence_sha256"),
        ):
            if not _SHA256.fullmatch(value):
                raise ValueError(f"invalid {name}")
        if not _ID.fullmatch(self.event_id):
            raise ValueError("invalid event_id")
        if self.tick < 0 or self.sequence < 0:
            raise ValueError("tick and sequence must be non-negative integers")
        if not self.producer_identity.strip():
            raise ValueError("producer_identity must not be empty")
        if self.lane in NON_CANONICAL_LANES and self.canonical:
            raise ValueError("cognitive side-channel output cannot be canonical")
        if self.lane == Lane.MOTOR_AUTHORIZATION and not self.canonical:
            raise ValueError("motor authorization must be canonical")

    def canonical_record(self) -> Mapping[str, Any]:
        return {
            "canonical": self.canonical,
            "causalParentSha256": self.causal_parent_sha256,
            "eventId": self.event_id,
            "lane": self.lane.value,
            "payloadSha256": self.payload_sha256,
            "policySha256": self.policy_sha256,
            "previousEvidenceSha256": self.previous_evidence_sha256,
            "producerIdentity": self.producer_identity,
            "revisionSha": self.revision_sha,
            "schemaVersion": self.schema_version,
            "sequence": self.sequence,
            "sideChannelReference": self.side_channel_reference,
            "systemId": self.system_id,
            "tick": self.tick,
        }

    def evidence_sha256(self) -> str:
        encoded = json.dumps(
            self.canonical_record(),
            ensure_ascii=False,
            separators=(",", ":"),
            sort_keys=True,
        ).encode("utf-8")
        return hashlib.sha256(encoded).hexdigest()


def verify_lane_transition(source: Lane, target: Lane) -> bool:
    allowed: Mapping[Lane, frozenset[Lane]] = {
        Lane.SENSORY_INTAKE: frozenset({Lane.THALAMIC_ROUTING, Lane.QUARANTINE}),
        Lane.THALAMIC_ROUTING: frozenset({
            Lane.REFLEX_SAFETY,
            Lane.DETERMINISTIC_VERIFICATION,
            Lane.COGNITIVE_SIDE_CHANNEL,
            Lane.QUARANTINE,
        }),
        Lane.REFLEX_SAFETY: frozenset({Lane.EVIDENCE, Lane.MOTOR_AUTHORIZATION, Lane.QUARANTINE}),
        Lane.DETERMINISTIC_VERIFICATION: frozenset({Lane.EVIDENCE, Lane.MOTOR_AUTHORIZATION, Lane.QUARANTINE}),
        Lane.COGNITIVE_SIDE_CHANNEL: frozenset({Lane.EVIDENCE}),
        Lane.EVIDENCE: frozenset({Lane.PERSISTENCE, Lane.CEREBELLAR_CORRECTION}),
        Lane.PERSISTENCE: frozenset({Lane.CEREBELLAR_CORRECTION}),
        Lane.CEREBELLAR_CORRECTION: frozenset({Lane.EVIDENCE, Lane.QUARANTINE}),
        Lane.MOTOR_AUTHORIZATION: frozenset({Lane.EVIDENCE, Lane.CEREBELLAR_CORRECTION}),
        Lane.HOMEOSTASIS: frozenset({Lane.THALAMIC_ROUTING, Lane.QUARANTINE}),
        Lane.QUARANTINE: frozenset({Lane.EVIDENCE}),
    }
    return target in allowed[source]


def validate_causal_chain(envelopes: Sequence[EvidenceEnvelope]) -> tuple[bool, str]:
    if not envelopes:
        return False, "EMPTY_CHAIN"
    previous_hash = envelopes[0].previous_evidence_sha256
    previous_sequence = envelopes[0].sequence - 1
    previous_tick = envelopes[0].tick
    for envelope in envelopes:
        if envelope.previous_evidence_sha256 != previous_hash:
            return False, "PREVIOUS_HASH_MISMATCH"
        if envelope.sequence != previous_sequence + 1:
            return False, "SEQUENCE_GAP"
        if envelope.tick < previous_tick:
            return False, "TICK_REGRESSION"
        previous_hash = envelope.evidence_sha256()
        previous_sequence = envelope.sequence
        previous_tick = envelope.tick
    return True, "VERIFIED"

File: src/features/product/runtime/neuroArchitectureContract.ts
export const NEURO_ARCHITECTURE_SCHEMA_VERSION =
  "sovereign.neuro-architecture-envelope.v1" as const;

export type EvidenceClass = "E0" | "E1" | "E2" | "E3" | "E4";

export type NeuroLane =
  | "sensory-intake"
  | "thalamic-routing"
  | "reflex-safety"
  | "deterministic-verification"
  | "cognitive-side-channel"
  | "evidence"
  | "persistence"
  | "cerebellar-correction"
  | "motor-authorization"
  | "homeostasis"
  | "quarantine";

export interface NeuroAliasBinding {
  readonly alias: string;
  readonly canonicalComponent: string;
  readonly evidenceClass: EvidenceClass;
  readonly softwareScope: "sovereign-studio-ato" | "areloria-wasd-side-channel";
  readonly scientificClaim: string;
  readonly limitations: readonly string[];
}

export interface NeuroEvidenceEnvelope {
  readonly schemaVersion: typeof NEURO_ARCHITECTURE_SCHEMA_VERSION;
  readonly systemId: string;
  readonly revisionSha: string;
  readonly policySha256: string;
  readonly eventId: string;
  readonly lane: NeuroLane;
  readonly tick: string;
  readonly sequence: string;
  readonly payloadSha256: string;
  readonly causalParentSha256: string;
  readonly previousEvidenceSha256: string;
  readonly producerIdentity: string;
  readonly canonical: boolean;
  readonly sideChannelReference?: string;
}

export interface ValidationResult {
  readonly ok: boolean;
  readonly errors: readonly string[];
}

const SHA40 = /^[0-9a-f]{40}$/;
const SHA256 = /^[0-9a-f]{64}$/;
const STABLE_ID = /^[a-z0-9][a-z0-9._:-]{1,159}$/;
const UINT = /^(0|[1-9][0-9]*)$/;

export const canonicalTruthLanes: ReadonlySet<NeuroLane> = new Set([
  "deterministic-verification",
  "evidence",
  "motor-authorization",
]);

export const nonCanonicalLanes: ReadonlySet<NeuroLane> = new Set([
  "cognitive-side-channel",
]);

const allowedTransitions: Readonly<Record<NeuroLane, ReadonlySet<NeuroLane>>> = {
  "sensory-intake": new Set(["thalamic-routing", "quarantine"]),
  "thalamic-routing": new Set([
    "reflex-safety",
    "deterministic-verification",
    "cognitive-side-channel",
    "quarantine",
  ]),
  "reflex-safety": new Set(["evidence", "motor-authorization", "quarantine"]),
  "deterministic-verification": new Set(["evidence", "motor-authorization", "quarantine"]),
  "cognitive-side-channel": new Set(["evidence"]),
  evidence: new Set(["persistence", "cerebellar-correction"]),
  persistence: new Set(["cerebellar-correction"]),
  "cerebellar-correction": new Set(["evidence", "quarantine"]),
  "motor-authorization": new Set(["evidence", "cerebellar-correction"]),
  homeostasis: new Set(["thalamic-routing", "quarantine"]),
  quarantine: new Set(["evidence"]),
};

export function verifyLaneTransition(source: NeuroLane, target: NeuroLane): boolean {
  return allowedTransitions[source].has(target);
}

export function validateNeuroAliasBinding(binding: NeuroAliasBinding): ValidationResult {
  const errors: string[] = [];
  if (!binding.alias.trim()) errors.push("EMPTY_ALIAS");
  if (!STABLE_ID.test(binding.canonicalComponent)) errors.push("INVALID_CANONICAL_COMPONENT");
  if (!binding.scientificClaim.trim()) errors.push("EMPTY_SCIENTIFIC_CLAIM");
  if (
    binding.softwareScope !== "sovereign-studio-ato" &&
    binding.softwareScope !== "areloria-wasd-side-channel"
  ) {
    errors.push("PROJECT_ISOLATION_VIOLATION");
  }
  return { ok: errors.length === 0, errors };
}

export function validateNeuroEvidenceEnvelope(
  envelope: NeuroEvidenceEnvelope,
): ValidationResult {
  const errors: string[] = [];
  if (envelope.schemaVersion !== NEURO_ARCHITECTURE_SCHEMA_VERSION) {
    errors.push("UNSUPPORTED_SCHEMA_VERSION");
  }
  if (!STABLE_ID.test(envelope.systemId)) errors.push("INVALID_SYSTEM_ID");
  if (!STABLE_ID.test(envelope.eventId)) errors.push("INVALID_EVENT_ID");
  if (!SHA40.test(envelope.revisionSha)) errors.push("INVALID_REVISION_SHA");
  if (!SHA256.test(envelope.policySha256)) errors.push("INVALID_POLICY_SHA256");
  if (!SHA256.test(envelope.payloadSha256)) errors.push("INVALID_PAYLOAD_SHA256");
  if (!SHA256.test(envelope.causalParentSha256)) errors.push("INVALID_CAUSAL_PARENT_SHA256");
  if (!SHA256.test(envelope.previousEvidenceSha256)) {
    errors.push("INVALID_PREVIOUS_EVIDENCE_SHA256");
  }
  if (!UINT.test(envelope.tick)) errors.push("INVALID_TICK");
  if (!UINT.test(envelope.sequence)) errors.push("INVALID_SEQUENCE");
  if (!envelope.producerIdentity.trim()) errors.push("EMPTY_PRODUCER_IDENTITY");
  if (nonCanonicalLanes.has(envelope.lane) && envelope.canonical) {
    errors.push("SIDE_CHANNEL_CANNOT_BE_CANONICAL");
  }
  if (envelope.lane === "motor-authorization" && !envelope.canonical) {
    errors.push("MOTOR_AUTHORIZATION_MUST_BE_CANONICAL");
  }
  return { ok: errors.length === 0, errors };
}

export const defaultNeuroAliasBindings: readonly NeuroAliasBinding[] = [
  {
    alias: "Thalamus",
    canonicalComponent: "neuro.thalamic-router",
    evidenceClass: "E1",
    softwareScope: "sovereign-studio-ato",
    scientificClaim:
      "Thalamic nuclei relay and regulate sensory, motor, limbic, arousal and cortical communication rather than acting as a single undifferentiated gateway.",
    limitations: ["Software routing is a functional analogy, not a biological replica."],
  },
  {
    alias: "Hypothalamus",
    canonicalComponent: "neuro.homeostasis-controller",
    evidenceClass: "E1",
    softwareScope: "sovereign-studio-ato",
    scientificClaim:
      "Hypothalamic nuclei coordinate autonomic, endocrine and homeostatic regulation through distributed connections.",
    limitations: ["Metrics and rate limiting model regulation only; they do not model endocrine biology."],
  },
  {
    alias: "Cerebellum",
    canonicalComponent: "neuro.execution-correction",
    evidenceClass: "E1",
    softwareScope: "sovereign-studio-ato",
    scientificClaim:
      "Cerebellar circuits compare intended movement with feedback and adjust timing, force and trajectory.",
    limitations: ["The software lane compares planned, executed and observed outcomes only."],
  },
  {
    alias: "Hippocampus",
    canonicalComponent: "neuro.evidence-consolidation",
    evidenceClass: "E1",
    softwareScope: "sovereign-studio-ato",
    scientificClaim:
      "Hippocampal and medial temporal systems contribute to memory registration, consolidation and retrieval through wider limbic and cortical connections.",
    limitations: ["Evidence persistence is not equivalent to human episodic memory."],
  },
  {
    alias: "Amygdala",
    canonicalComponent: "neuro.reversible-safety-reflex",
    evidenceClass: "E1",
    softwareScope: "sovereign-studio-ato",
    scientificClaim:
      "Amygdaloid nuclei participate in salience, stress-related, memory and behavioral regulation through cortical, hypothalamic, thalamic and striatal circuits.",
    limitations: ["The lane may trigger reversible protection only, never an irreversible effect."],
  },
];
